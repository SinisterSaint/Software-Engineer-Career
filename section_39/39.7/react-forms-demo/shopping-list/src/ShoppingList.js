import React, { useState } from "react";
import NewListItemForm from "./NewListItemForm";
import { v4 as uuid } from "uuid";


function ShoppingList() {
  const [items, setItems] = useState([]);

  const renderItems = () => {
    return (
      <ul>
        {items.map(item => (
          <li key={item.id}>
            {item.name}: {item.qty}
          </li>
        ))}
      </ul>
    );
  };
  // end renderItems

  /** Add new item object to cart. */
  const addItem = item => {
    let newItem = { ...item, id: uuid() };
    setItems(items => [...items, newItem]);
  };
  // end addItem

  return (
    <div className="ShoppingList">
      <NewListItemForm addItem={addItem} />
      {renderItems()}
    </div>
  );
};
// end

export default ShoppingList;
