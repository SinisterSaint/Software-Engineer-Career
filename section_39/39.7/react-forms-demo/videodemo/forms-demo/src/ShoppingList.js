import React, { useState } from "react";
import { v4 as uuid } from 'uuid';
import Item from "./Item"
import NewItemForm from "./NewItemForm";


const ShoppingList = () => {
  const INITIAL_STATE = [
    { id: uuid(), name: 'Peanut Butter', qty: 2 },
    { id: uuid(), name: 'Whole Milk', qty: 1 }
  ]
  const [items, setItems] = useState(INITIAL_STATE);
  const addItem = (newItem) => {
    setItems(items => [...items, { ...newItem, id: uuid() }])
  }
  return (
    <div>
      <h3>Shopping List</h3>
      <NewItemForm addItem={addItem} />
      <div>
        {items.map(({ id, name, qty }) => <Item id={id} name={name} qty={qty} key={id} />)}
      </div>

    </div>
  )

}

export default ShoppingList;


