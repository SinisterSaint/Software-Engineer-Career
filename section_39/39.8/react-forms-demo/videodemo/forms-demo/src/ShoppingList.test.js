import React from "react";
import { render, fireEvent } from "@testing-library/react";
import ShoppingList from "./ShoppingList";

it("renders without crashing", function () {
  render(<ShoppingList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<ShoppingList />);
  expect(asFragment()).toMatchSnapshot();
});

it("should add new item", function () {
  const { queryByText, getByLabelText } = render(<ShoppingList />);
  const input = getByLabelText("Product");
  const btn = queryByText("Add Item");
  expect(queryByText('Product Name: Chocolate Milk')).not.toBeInTheDocument();
  fireEvent.change(input, { target: { value: 'Chocolate Milk' } });
  fireEvent.click(btn);
  expect(queryByText('Product Name: Chocolate Milk')).toBeInTheDocument();
})


