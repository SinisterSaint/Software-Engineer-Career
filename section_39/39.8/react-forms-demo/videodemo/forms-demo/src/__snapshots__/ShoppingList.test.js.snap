// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3>
      Shopping List
    </h3>
    <form>
      <label
        for="name"
      >
        Product
      </label>
      <input
        id="name"
        name="name"
        placeholder="Item Name"
        type="text"
        value=""
      />
      <label
        for="qty"
      >
        Quantity:
      </label>
      <input
        id="qty"
        max="10"
        min="1"
        name="qty"
        type="range"
        value=""
      />
      <button>
        Add Item
      </button>
    </form>
    <div>
      <ul>
        <li>
          Product Name: Peanut Butter
        </li>
        <li>
          Quantity: 2
        </li>
      </ul>
      <ul>
        <li>
          Product Name: Whole Milk
        </li>
        <li>
          Quantity: 1
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
